src/Markup/Avalonia.Markup.Xaml.Loader/CompilerExtensions/GroupTransformers/XamlIncludeGroupTransformer.cs

```diff
@@ -0,0 +1,133 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using Avalonia.Markup.Xaml.XamlIl.CompilerExtensions.Transformers;
+using Avalonia.Platform;
+using XamlX;
+using XamlX.Ast;
+using XamlX.Emit;
+using XamlX.IL;
+using XamlX.Transform;
+using XamlX.Transform.Transformers;
+using XamlX.TypeSystem;
+
+namespace Avalonia.Markup.Xaml.XamlIl.CompilerExtensions.GroupTransformers;
+
+internal class AvaloniaXamlIncludeTransformer : IXamlAstGroupTransformer
+{
+    public IXamlAstNode Transform(AstGroupTransformationContext context, IXamlAstNode node)
+    {
+        if (node is not XamlValueWithManipulationNode valueNode
+            || valueNode.Value is not XamlAstNewClrObjectNode objectNode
+            || (objectNode.Type.GetClrType() != context.GetAvaloniaTypes().StyleInclude
+                && objectNode.Type.GetClrType() != context.GetAvaloniaTypes().ResourceInclude))
+        {
+            return node;
+        }
+
+        var nodeTypeName = objectNode.Type.GetClrType().Name;
+        if (valueNode.Manipulation is not XamlObjectInitializationNode
+            {
+                Manipulation: XamlPropertyAssignmentNode { Property: { Name: "Source" } } sourceProperty
+            })
+        {
+            return context.ParseError($"Source property must be set on the \"{nodeTypeName}\" node.", node);
+        }
+
+        if (sourceProperty.Values.OfType<XamlAstNewClrObjectNode>().FirstOrDefault() is not { } sourceUriNode
+            || sourceUriNode.Type.GetClrType() != context.GetAvaloniaTypes().Uri
+            || sourceUriNode.Arguments.FirstOrDefault() is not XamlConstantNode { Constant: string originalAssetPath })
+        {
+            // TODO: make it a compiler warning
+            // Source value can be set with markup extension instead of the Uri object node, we don't support it here yet.
+            return node;
+        }
+
+        if (originalAssetPath.StartsWith("avares://"))
+        {
+        }
+        else if (originalAssetPath.StartsWith("/"))
+        {
+            var baseUrl = context.CurrentDocument.Uri ?? throw new InvalidOperationException("CurrentDocument URI is null.");
+            originalAssetPath = baseUrl.Substring(0, baseUrl.LastIndexOf('/')) + originalAssetPath;
+        }
+        else
+        {
+            return context.ParseError(
+                $"Avalonia supports only \"avares://\" sources or relative sources starting with \"/\" on the \"{nodeTypeName}\" node.",
+                node);
+        }
+        
+        AssetLoader.RegisterResUriParsers();
```

Comment:
I believe this should happen somewhere in MSBuild task initialization rather than in a transformer. I'm also not quite sure if that's a good idea to pollute MSBuild process with our global state

Reply:
I will move. Also I don't use avalonia global state here, don't access AssetLoader APIs. Instead only URI scheme configuration is registered in .NET URI parser. Otherwise it will lower case all the links.